core/domain/expression_parser.py

```diff
@@ -117,10 +117,22 @@ def tokenize(expression):
     """
     expression = expression.replace(' ', '')
 
-    re_string = r'([a-zA-Z]+|[0-9]+\.[0-9]+|[0-9]+|[%s])' % '\\'.join(
-        _VALID_OPERATORS)
+    # Note: Greek letters and math functions are sorted in reverse by length so
+    # that longer ones get matched before shorter ones.
+    # For eg. 'x + epsilon' should be tokenized as ['x','+','epsilon'] and not
+    # ['x','+','e','*','psi','*','l','*','o','*','n'].
+    re_string = r'(%s|[a-zA-Z]|[0-9]+\.[0-9]+|[0-9]+|[%s])' % (
+        '|'.join(sorted(
+            _GREEK_LETTERS + _MATH_FUNCTION_NAMES, reverse=True, key=len)),
+        '\\'.join(_VALID_OPERATORS))
 
     token_texts = re.findall(re_string, expression)
+
+    if sum([len(token_text) for token_text in token_texts]) != len(
+            expression):
+        raise Exception(
+            'Invalid syntax: Unexpected token present in the expression.')
```

Comment:
Which one? Give more explicit error messages. They are needed for debugging.

Reply:
Done.